File: python/03_map_filter_reduce/map_filter_reduce.py
"""
MAP / FILTER / REDUCE
=====================
Las tres funciones de orden superior más usadas en programación funcional
para transformar colecciones de forma declarativa.

  map(f, colección)    → aplica f a cada elemento
  filter(pred, col)    → conserva elementos donde pred es True
  reduce(f, col)       → combina todos los elementos en uno
"""

from functools import reduce

numeros = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

# -----------------------------------------------
# MAP: transformar cada elemento
# -----------------------------------------------
cuadrados      = list(map(lambda x: x ** 2, numeros))
cuadrados_gen  = [x ** 2 for x in numeros]   # list comprehension equivalente

# -----------------------------------------------
# FILTER: seleccionar elementos
# -----------------------------------------------
pares    = list(filter(lambda x: x % 2 == 0, numeros))
pares_lc = [x for x in numeros if x % 2 == 0]

# -----------------------------------------------
# REDUCE: acumular / colapsar
# -----------------------------------------------
suma_total   = reduce(lambda acc, x: acc + x, numeros, 0)
producto     = reduce(lambda acc, x: acc * x, numeros, 1)
maximo       = reduce(lambda acc, x: x if x > acc else acc, numeros)

# -----------------------------------------------
# Encadenamiento: pipeline funcional
# -----------------------------------------------
# "La suma de los cuadrados de los números impares"
resultado = reduce(
    lambda acc, x: acc + x,
    map(lambda x: x ** 2,
        filter(lambda x: x % 2 != 0, numeros)),
    0
)

# Con list comprehension equivalente:
resultado_lc = sum(x ** 2 for x in numeros if x % 2 != 0)


# -----------------------------------------------
# Ejemplo práctico: procesar lista de productos
# -----------------------------------------------
productos = [
    {"nombre": "Laptop",  "precio": 1200, "categoria": "tecnologia"},
    {"nombre": "Libro",   "precio": 25,   "categoria": "educacion"},
    {"nombre": "Monitor", "precio": 350,  "categoria": "tecnologia"},
    {"nombre": "Silla",   "precio": 180,  "categoria": "muebles"},
    {"nombre": "Teclado", "precio": 80,   "categoria": "tecnologia"},
]

# Solo tecnología con precio < 500, mostrar nombres en mayúsculas
nombres_tech_baratos = list(map(
    lambda p: p["nombre"].upper(),
    filter(lambda p: p["categoria"] == "tecnologia" and p["precio"] < 500, productos)
))

total_tecnologia = reduce(
    lambda acc, p: acc + p["precio"],
    filter(lambda p: p["categoria"] == "tecnologia", productos),
    0
)

# -----------------------------------------------
# DEMO
# -----------------------------------------------
if __name__ == "__main__":
    print("=== map ===")
    print(f"Cuadrados: {cuadrados}")

    print("\n=== filter ===")
    print(f"Pares: {pares}")

    print("\n=== reduce ===")
    print(f"Suma: {suma_total}")
    print(f"Producto: {producto}")
    print(f"Máximo: {maximo}")

    print("\n=== Pipeline encadenado ===")
    print(f"Suma de cuadrados de impares: {resultado}")
    print(f"Equivalente con comprehension: {resultado_lc}")

    print("\n=== Ejemplo productos ===")
    print(f"Tech barata (< $500): {nombres_tech_baratos}")
    print(f"Total inversión tecnología: ${total_tecnologia}")

File: java/src/main/java/com/paradigmas/fp/streams/MapFilterReduce.java
package com.paradigmas.fp.streams;

import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

/**
 * MAP / FILTER / REDUCE con la Stream API de Java
 * =================================================
 * Stream.map()     → transforma cada elemento
 * Stream.filter()  → selecciona elementos según predicado
 * Stream.reduce()  → acumula / colapsa a un solo valor
 * Stream.collect() → materializa el stream en una colección
 *
 * Java 26: var, records, pattern matching, toList()
 */
public class MapFilterReduce {

    record Producto(String nombre, double precio, String categoria) {}

    public static void main(String[] args) {

        var numeros = List.of(1, 2, 3, 4, 5, 6, 7, 8, 9, 10);

        // ── MAP ──────────────────────────────────────────────────────
        System.out.println("=== map ===");
        var cuadrados = numeros.stream()
                               .map(x -> x * x)
                               .toList();
        System.out.println("Cuadrados: " + cuadrados);

        var cadenas = numeros.stream()
                             .map(x -> "num_" + x)
                             .toList();
        System.out.println("Cadenas:   " + cadenas);

        // ── FILTER ───────────────────────────────────────────────────
        System.out.println("\n=== filter ===");
        var pares = numeros.stream()
                           .filter(x -> x % 2 == 0)
                           .toList();
        System.out.println("Pares:    " + pares);

        var mayoresDe5 = numeros.stream()
                                .filter(x -> x > 5)
                                .toList();
        System.out.println("Mayores de 5: " + mayoresDe5);

        // ── REDUCE ───────────────────────────────────────────────────
        System.out.println("\n=== reduce ===");
        int suma = numeros.stream()
                          .reduce(0, Integer::sum);
        System.out.println("Suma:     " + suma);

        int producto = numeros.stream()
                              .reduce(1, (acc, x) -> acc * x);
        System.out.println("Producto: " + producto);

        Optional<Integer> maximo = numeros.stream()
                                          .reduce(Integer::max);
        System.out.println("Máximo:   " + maximo.orElse(-1));

        // ── PIPELINE encadenado ──────────────────────────────────────
        System.out.println("\n=== Pipeline encadenado ===");
        // Suma de cuadrados de números impares
        int sumaCuadImpares = numeros.stream()
                                     .filter(x -> x % 2 != 0)
                                     .map(x -> x * x)
                                     .reduce(0, Integer::sum);
        System.out.println("Suma de cuadrados de impares: " + sumaCuadImpares);

        // ── EJEMPLO PRÁCTICO: productos ──────────────────────────────
        System.out.println("\n=== Productos ===");
        var productos = List.of(
            new Producto("Laptop",  1200, "tecnologia"),
            new Producto("Libro",   25,   "educacion"),
            new Producto("Monitor", 350,  "tecnologia"),
            new Producto("Silla",   180,  "muebles"),
            new Producto("Teclado", 80,   "tecnologia")
        );

        // Nombres en mayúsculas de tech con precio < 500
        var techBarata = productos.stream()
                                  .filter(p -> p.categoria().equals("tecnologia"))
                                  .filter(p -> p.precio() < 500)
                                  .map(p -> p.nombre().toUpperCase())
                                  .toList();
        System.out.println("Tech barata (< $500): " + techBarata);

        // Total invertido en tecnología
        double totalTech = productos.stream()
                                    .filter(p -> p.categoria().equals("tecnologia"))
                                    .mapToDouble(Producto::precio)
                                    .sum();
        System.out.println("Total tecnología: $" + totalTech);

        // Agrupar por categoría
        Map<String, List<Producto>> porCategoria = productos.stream()
                                                            .collect(Collectors.groupingBy(Producto::categoria));
        System.out.println("\nProductos por categoría:");
        porCategoria.forEach((cat, prods) ->
            System.out.println("  " + cat + ": " + prods.stream().map(Producto::nombre).toList())
        );

        // Estadísticas de precios
        var stats = productos.stream()
                             .mapToDouble(Producto::precio)
                             .summaryStatistics();
        System.out.printf("%nEstadísticas precios → min: $%.0f, max: $%.0f, avg: $%.1f%n",
                          stats.getMin(), stats.getMax(), stats.getAverage());
    }
}
